go/vt/vtctl/grpcvtctldserver/server.go

```diff
@@ -2835,6 +2856,230 @@ func (s *VtctldServer) ValidateKeyspace(ctx context.Context, req *vtctldatapb.Va
+				errMessage := fmt.Sprintf("GetShard(%v, %v) failed: %v", keyspace, shard, err)
+				shardResp.Results = append(shardResp.Results, errMessage)
+				resp.Results = append(resp.Results, errMessage)
+				resp.ResultsByShard[shard] = &shardResp
+				return
+			}
+
+			if !si.HasPrimary() {
+				if !req.SkipNoPrimary {
+					errMessage := fmt.Sprintf("no primary in shard %v/%v", keyspace, shard)
+					shardResp.Results = append(shardResp.Results, errMessage)
+					resp.Results = append(resp.Results, errMessage)
+					resp.ResultsByShard[shard] = &shardResp
+				}
+				return
+			}
+
+			if referenceSchema == nil {
+				referenceAlias = si.PrimaryAlias
+				referenceSchema, err = schematools.GetSchema(ctx, s.ts, s.tmc, referenceAlias, nil, []string{} /*excludeTables*/, false /*includeViews*/)
+				if err != nil {
+					errMessage := fmt.Sprintf("GetSchema(%v, nil, %v, %v) failed: %v", referenceAlias, []string{} /*excludeTables*/, false /*includeViews*/, err)
+					shardResp.Results = append(shardResp.Results, errMessage)
+					resp.Results = append(resp.Results, errMessage)
+					resp.ResultsByShard[shard] = &shardResp
+					return
+				}
+			}
+
+			aliases, err := s.ts.FindAllTabletAliasesInShard(ctx, keyspace, shard)
+			if err != nil {
+				errMessage := fmt.Sprintf("FindAllTabletAliasesInShard(%v, %v) failed: %v", keyspace, shard, err)
+				shardResp.Results = append(shardResp.Results, errMessage)
+				resp.Results = append(resp.Results, errMessage)
+				resp.ResultsByShard[shard] = &shardResp
+				return
+			}
+
+			aliasWg := sync.WaitGroup{}
+			aliasErrs := concurrency.AllErrorRecorder{}
+
+			for _, alias := range aliases {
+				// Don't diff schemas for self
+				if referenceAlias == alias {
+					continue
+				}
+				aliasWg.Add(1)
+				go func(alias *topodatapb.TabletAlias) {
+					defer aliasWg.Done()
+					replicaSchema, err := schematools.GetSchema(ctx, s.ts, s.tmc, alias, nil, req.ExcludeTables, req.IncludeViews)
+					if err != nil {
+						aliasErrs.RecordError(fmt.Errorf("GetSchema(%v, nil, %v, %v) failed: %v", alias, req.ExcludeTables, req.IncludeViews, err))
+						return
+					}
+
+					tmutils.DiffSchema(topoproto.TabletAliasString(referenceAlias), referenceSchema, topoproto.TabletAliasString(alias), replicaSchema, &aliasErrs)
+				}(alias)
+			}
+			aliasWg.Wait()
+
+			if aliasErrs.HasErrors() {
+				for _, err := range aliasErrs.Errors {
+					errMessage := err.Error()
+					shardResp.Results = append(shardResp.Results, errMessage)
+					resp.Results = append(resp.Results, errMessage)
+				}
+			}
+
+			resp.ResultsByShard[shard] = &shardResp
+		}(shard)
+	}
+
+	wg.Wait()
+
+	return &resp, nil
+}
+
+// ValidateVersionKeyspace validates all versions are the same in all
+// tablets in a keyspace
+func (s *VtctldServer) ValidateVersionKeyspace(ctx context.Context, req *vtctldatapb.ValidateVersionKeyspaceRequest) (*vtctldatapb.ValidateVersionKeyspaceResponse, error) {
+	keyspace := req.Keyspace
+	// find all the shards
```

Comment:
fwiw, i think a lot of these comments you copied over are superfluous and can be removed